### DISCUSSION TITLE
STM32F4(01) USB HID - Any work being done?

### DISCUSSION THREAD

**Discussion by Hyratel** [stm32duino/Arduino_Core_STM32#1644] [Libraries]
I picked up some 32F401 boards because they were superior in all specs to the ATmega32u4 (Leonardo et al) and given current shortaage, cheaper by about half, but discovered to my immense dismay that there seemed to be no work being done to bring USB HID to the chip. The option exists in CubeIDE (but CubeIDE is A Mess) for the chip to have USB HID Device enabled, but I can't find anything that plugs into STM32duino for the F4 (USBComposite library is exclusive to the F103, as best I can tell)

**Comment by fpistm** [comment]
Hi @Hyratel 
it seems you used this core: https://github.com/rogerclarkmelbourne/Arduino_STM32
Which integrate the USB composite library.

This one provide HID support in a different way and can be enabled in the menu.

**Reply by Hyratel** [reply]
I did see the "HID Mouse and keyboard" under "USB Support" in the menus but I haven't been able to find any information on what that does or how to interface with it. Specifically, I want to do work with USB Joysticks, and if at all possible, Force Feedback (which is its own box of rocks on top of HID).

(I already have one USB joystick with a Pro Micro under my belt, MHeironymous lib is Excellent)

**Comment by fpistm** [comment]
The STM32 core simply provides the default implementation of the Keyboard and Mouse form the Arduino standard library:
https://www.arduino.cc/reference/en/language/functions/usb/keyboard/
https://www.arduino.cc/reference/en/libraries/mouse/

You can use the default examples provided with Arduino IDE in `9. USB` submenu

**Reply by Hyratel** [reply]
the Joystick Library or a Flexible raw-HID frontend needs to be brought out; I've been bashing my head against it for weeks and trying to splice together the difference between arduino C++ Style and the Object-Oriented C used by the Core has me stymied.

**Reply by elMariachi90** [reply]
Did you manage to get anywhere? Having the same issue currently, I would love to use my F401 as a custom Gaming Peripheral with custom HID over USB but there seems to be no information around

**Reply by Hyratel** [reply]
look at https://github.com/Levi--G/USBLibrarySTM32
